### PR TITLE
[Deprecated] Add Swish operator (replaced by #7172)

### LINKED ISSUE #5853: Request for Swish Op
# Swish/SiLU
Do you have any plans to implement the Swish Op in ONNX?

### Describe the operator
Swish is a popular Activation fuction. Its mathematical definition could be found at https://en.wikipedia.org/wiki/Swish_function

TensorFLow has https://www.tensorflow.org/api_docs/python/tf/nn/silu
Keras has https://keras.io/api/layers/activations/ (also in https://www.tensorflow.org/api_docs/python/tf/keras/activations/swish)

Pytorch has https://pytorch.org/docs/stable/generated/torch.nn.SiLU.html 

### Can this operator be constructed using existing onnx operators?
Yes, it could be implemented as a combination of Mul and Sigmoid Ops:
 x * Sigmoid (beta * x)

### Is this operator used by any model currently? Which one?
Yes. Modern Yolo series like yolov5, yolov7, yolov8, yolop and EfficientNet all have such Swish Ops.  

Yolov5: https://github.com/ultralytics/yolov5/blob/master/models/tf.py#L224 

EfficientNet:
https://paperswithcode.com/method/efficientnet which has Swish in https://github.com/lukemelas/EfficientNet-PyTorch/blob/2eb7a7d264344ddf15d0a06ee99b0dca524c6a07/efficientnet_pytorch/model.py#L294 

### Are you willing to contribute it? (Y/N)
Possibly Yes.

### PR DESCRIPTION
**Pull request by isdanni**
### Description
These changes have been made to support the SiLU/Swish operator as a function op.

### Motivation and Context
Close #5853 

The SiLU function can be defined as $x * Sigmoid (alpha * x)$ where $alpha = 1.0$.

Swish ops is used in Yolo series and EfficientNet.

### CODE REVIEW

**Comment by isdanni** [comment]
Would appreciate some guidance on adding tests to `backend/test/data/node` :smiley:

File: onnx/defs/math/defs.cc [hunk 1]
```diff
@@ -644,6 +644,25 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SiLU,
+    21,
+    OpSchema()
+        .SetDoc(SiLU_ver21_doc)
+        .Input(0, "X", "Input tensor", "T", OpSchema::Single, true, 1, OpSchema::Differentiable)
+        .Output(0, "Y", "Output tensor", "T", OpSchema::Single, true, 1, OpSchema::Differentiable)
+        .TypeConstraint(
```
[the remark this answers is not in the document]
Comment (isdanni) [reply onnx/defs/math/defs.cc:660]: If we are using SiLU I'm assuming beta is 1.0? I can add it tho

File: onnx/defs/math/defs.cc [hunk 2]
```diff
@@ -644,6 +644,25 @@ ONNX_OPERATOR_SET_SCHEMA(
         .SetContextDependentFunctionBodyBuilder(BuildContextDependentFunctionBodyGelu)
         .TypeAndShapeInferenceFunction(propagateShapeAndTypeFromFirstInput));
 
+static const char* SiLU_ver21_doc = R"DOC(
+Sigmoid Linear Unit (SiLU), or known as the Swish function, takes one input data (Tensor<T>)
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:648]: Since we have `HardSwish`, would it be more uniform to call this function-op `Swish` ? Just a minor matter, just wondering.

File: onnx/defs/math/defs.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnx/defs/math/defs.cc:660]: I notice HardSwish uses alpha and beta in the form `alpha*x + beta` ?

**Comment by gramalingam** [comment]
> Would appreciate some guidance on adding tests to `backend/test/data/node` 😃

Please take a look at [this line](https://github.com/onnx/onnx/blob/d7e2b813c1a26cdb8bcb6c4547cb5e6ffeb9b49b/tools/update_doc.sh#L27) and [this line](https://github.com/onnx/onnx/blob/d7e2b813c1a26cdb8bcb6c4547cb5e6ffeb9b49b/tools/update_doc.sh#L30) for updating generated files.

File: onnx/defs/math/defs.cc (hunk 2 above)
Comment (isdanni) [reply onnx/defs/math/defs.cc:648]: good idea! I will update this to `swish` and add the beta parameter.

File: onnx/defs/math/defs.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (isdanni) [reply onnx/defs/math/defs.cc:660]: Based on https://github.com/onnx/onnx/pull/5964#discussion_r1511720346 I will add `beta` and rename the operator to `swish`, thanks for reviewing this!

File: onnx/defs/math/defs.cc [hunk 3]
```diff
@@ -644,6 +644,33 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+        .Output(0, "Y", "Output tensor", "T", OpSchema::Single, true, 1, OpSchema::Differentiable)
+        .TypeConstraint(
+          "T",
+            {"tensor(float16)", "tensor(float)", "tensor(bfloat16)"},
+            "Constrain input and output types to float tensors.")
+        .TypeAndShapeInferenceFunction(propagateShapeAndTypeFromFirstInput)
+        .FunctionBody(
+            R"ONNX(
+          {
+            S_X = Sigmoid<beta = 1.0>(X)
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:668]: Sigmoid doesn't have an attribute.  We should be doing something like below:
```py
   Beta = Constant <value_float: float = @beta>()
   BetaCast = CastLike (Beta, X)
   BetaX = Mul (BetaCast, X)
   SigmoidBetaX = Sigmoid(BetaX)
   Y = Mul (X, SigmoidBetaX)
```

Comment (gramalingam) [reply onnx/defs/math/defs.cc:668]: However, as commented above, it seems to me we should be calling this `alpha` instead of `beta`, based on the convention in `HardSwish`.

File: onnx/defs/math/defs.cc [hunk 4]
```diff
@@ -644,6 +644,33 @@ ONNX_OPERATOR_SET_SCHEMA(
...
         .TypeAndShapeInferenceFunction(propagateShapeAndTypeFromFirstInput));
 
+static const char* Swish_ver21_doc = R"DOC(
+Swish function takes one input data (Tensor<T>) and produces one output data (Tensor<T>) of the same shape,
+where $Swish(x) = x * sigmoid(beta * x)$.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Swish,
+    21,
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:654]: This will now need to be 22

File: onnx/test/version_converter/automatic_upgrade_test.py [hunk 5]
```diff
@@ -1198,6 +1198,9 @@ def test_Sigmoid(self) -> None:
     def test_Sign(self) -> None:
         self._test_op_upgrade("Sign", 9)
 
+    def test_Swish(self) -> None:
+        self._test_op_upgrade("Swish", 21)
```
Comment (gramalingam) [inline onnx/test/version_converter/automatic_upgrade_test.py:1202]: Now 22

**Comment by isdanni** [comment]
> > Would appreciate some guidance on adding tests to `backend/test/data/node` 😃
> 
> Please take a look at [this line](https://github.com/onnx/onnx/blob/d7e2b813c1a26cdb8bcb6c4547cb5e6ffeb9b49b/tools/update_doc.sh#L27) and [this line](https://github.com/onnx/onnx/blob/d7e2b813c1a26cdb8bcb6c4547cb5e6ffeb9b49b/tools/update_doc.sh#L30) for updating generated files.

Thanks for the review!! One followup question: when I run both commands after updating the operator to 22, there are ~20 `.onnx` files changed, instead of just a few `output_0.pb` in the last commit, is this normal behaviour?

File: onnx/defs/math/defs.cc [hunk 6]
```diff
@@ -621,6 +621,35 @@ ONNX_OPERATOR_SET_SCHEMA(
         .SetContextDependentFunctionBodyBuilder(BuildContextDependentFunctionBodyGelu)
         .TypeAndShapeInferenceFunction(propagateShapeAndTypeFromFirstInput));
 
+static const char* Swish_ver22_doc = R"DOC(
+Swish function takes one input data (Tensor<T>) and produces one output data (Tensor<T>) of the same shape,
+where $Swish(x) = x * sigmoid(beta * x)$.
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:626]: This needs to change to `alpha`

File: onnx/defs/math/defs.cc [hunk 7]
```diff
@@ -621,6 +621,35 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Swish,
+    22,
+    OpSchema()
+        .SetDoc(Swish_ver22_doc)
+        .Input(0, "X", "Input tensor", "T", OpSchema::Single, true, 1, OpSchema::Differentiable)
+        .Output(0, "Y", "Output tensor", "T", OpSchema::Single, true, 1, OpSchema::Differentiable)
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:635]: Where is the attribute `alpha`? It's definition should be added here.

**Comment by gramalingam** [comment]
> > > Would appreciate some guidance on adding tests to `backend/test/data/node` 😃
> > 
> > 
> > Please take a look at [this line](https://github.com/onnx/onnx/blob/d7e2b813c1a26cdb8bcb6c4547cb5e6ffeb9b49b/tools/update_doc.sh#L27) and [this line](https://github.com/onnx/onnx/blob/d7e2b813c1a26cdb8bcb6c4547cb5e6ffeb9b49b/tools/update_doc.sh#L30) for updating generated files.
> 
> Thanks for the review!! One followup question: when I run both commands after updating the operator to 22, there are ~20 `.onnx` files changed, instead of just a few `output_0.pb` in the last commit, is this normal behaviour?

Apologies for the delayed reply! Missed this one. Something seems to be wrong ... too many test files in the data folder. Wonder if it is some problem from merging the latest changes from main?

**Comment by gramalingam** [comment]
The code-freeze for version 1.17 is coming up ... we may need to move this to 1.18, if not completed by this month end.

**Comment by justinchuby** [comment]
@isdanni checking in, do you intend to move this PR forward?

**Comment by isdanni** [comment]
> @isdanni checking in, do you intend to move this PR forward?

Yes, I will rebase it and provide an update soon. Apologies for letting it go stale

**Comment by isdanni** [comment]
> @isdanni checking in, do you intend to move this PR forward?

I fixed the CI tests and the PR is ready for review now. Really sorry for the delay! 🙏

File: onnx/test/version_converter/automatic_upgrade_test.py (hunk 5 above)
Comment (andife) [reply onnx/test/version_converter/automatic_upgrade_test.py]: Now opset version 24? https://github.com/onnx/onnx/blob/main/docs/Versioning.md#released-versions

File: onnx/defs/math/defs.cc (hunk 4 above)
Comment (andife) [reply onnx/defs/math/defs.cc]: Now opset version 24? https://github.com/onnx/onnx/blob/main/docs/Versioning.md#released-versions

**Comment by andife** [comment]
@isdanni opset version would be 24 Now ? https://github.com/onnx/onnx/blob/main/docs/Versioning.md#released-versions

**Comment by justinchuby** [comment]
Feel free to make the PR as ready when it's ready for review, thanks!

File: onnx/defs/math/defs.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnx/defs/math/defs.cc:618]: Were the changes made?

[the remark this answers is not in the document]
Comment (isdanni) [reply onnx/defs/math/defs.cc:618]: yes, I used [alpha](https://github.com/onnx/onnx/pull/5964/files#diff-424d8e81bbc9ba3bef6dbd569c50893481011388605157350d5a672799c303f3R614) for scaling of the input before sigmoid and is default to `1.0`. tests can be found [here](https://github.com/onnx/onnx/pull/5964/files#diff-3d133fe118c467bb7aa22cb4f6a0b84dd073f31753f3f5658ae794e1724b6da3R28).

Function is defined as:

$x * Sigmoid (alpha * x)$ where $alpha = 1.0$.

File: onnx/defs/math/defs.cc [hunk 8]
```diff
@@ -602,6 +602,36 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            R"ONNX(
+            {
+                Alpha = Constant <value_float: float = @alpha>()
+                AlphaCast = CastLike (Alpha, X)
+                AlphaMulX = Mul (AlphaCast, X)
+                SigmoidAlphaMulX = Sigmoid(AlphaMulX)
+                Y = Mul (X, SigmoidAlphaMulX)
+            }
+            )ONNX",
+            18));
```
Comment (justinchuby) [inline onnx/defs/math/defs.cc]:
```suggestion
            )ONNX"));
```

**Comment by justinchuby** [comment]
I think we need a reference implementation and this will be ready to merge. Could you help add that?

**Comment by isdanni** [comment]
> I think we need a reference implementation and this will be ready to merge. Could you help add that?

sounds good! I will add it in `onnx/reference/ops/op_swish.py` tonight. 

(codebase is set up on my personal laptop.)

**Comment by justinchuby** [comment]
branch cut is end of week this week. It would be good to have this PR ready by then

File: onnx/defs/math/defs.cc [hunk 9]
```diff
@@ -602,6 +602,35 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+    Swish,
+    24,
+    OpSchema()
+        .Attr("alpha", "Coefficient to multiply with input before sigmoid.", AttributeProto::FLOAT, 1.0f)
+        .SetDoc(Swish_ver24_doc)
+        .Input(0, "X", "Input tensor", "T", OpSchema::Single, true, 1, OpSchema::Differentiable)
+        .Output(0, "Y", "Output tensor", "T", OpSchema::Single, true, 1, OpSchema::Differentiable)
+        .TypeConstraint(
+            "T",
+            {"tensor(float16)", "tensor(float)", "tensor(bfloat16)"},
```
Comment (cbourjau) [inline onnx/defs/math/defs.cc]:
```suggestion
            {"tensor(float16)", "tensor(float)", "tensor(bfloat16)", "tensor(double)"},
```
It would be very helpful if we had a general policy on which types to use by default. Not having `double` is annoying because there is no lossless workaround.

File: onnx/backend/test/data/node/test_acos/test_data_set_0/output_0.pb
Comment (cbourjau) [inline onnx/backend/test/data/node/test_acos/test_data_set_0/output_0.pb:1]: Why were all these seemingly unrelated files updated?

Comment (isdanni) [reply onnx/backend/test/data/node/test_acos/test_data_set_0/output_0.pb:1]: This is generated by running `python onnx/backend/test/cmd_tools.py generate-data`. Otherwise, it wouldn't pass the CI tests.

Comment (justinchuby) [inline onnx/backend/test/data/node/test_acos/test_data_set_0/output_0.pb:1]: I don’t think you need these files that is not swish related. Could you remove? We can see how CI reacts

File: onnx/reference/ops/op_swish.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+from __future__ import annotations
+
+import numpy as np
+
+from onnx.reference.ops._op import OpRunUnaryNum
+
+
+class Swish(OpRunUnaryNum):
+    def _run(self, x, alpha=None):
+        alpha = alpha or self.alpha
```
Comment (justinchuby) [inline onnx/reference/ops/op_swish.py]: Making sure: where is self.alpha defined?

File: onnx/backend/test/data/node/test_acos/test_data_set_0/output_0.pb
Comment (justinchuby) [reply onnx/backend/test/data/node/test_acos/test_data_set_0/output_0.pb:1]: The file difference are probably due to system/protobuf version differences. We shouldn’t need the rest of the files that are not unique to swish.  @isdanni

Comment (justinchuby) [reply onnx/backend/test/data/node/test_acos/test_data_set_0/output_0.pb:1]: @cbourjau this is also an issue with the binary proto files: they are not bitwise equivalent across environments

File: onnx/reference/ops/op_swish.py (hunk 10 above)
Comment (isdanni) [reply onnx/reference/ops/op_swish.py]: I actually followed the other ops like https://github.com/onnx/onnx/blob/e65124429fe849558e5ed4fd98e2f1fc19074caf/onnx/reference/ops/op_elu.py#L12, and from other reference ops attributes should be dynamically defined in OpRun: https://github.com/onnx/onnx/blob/35655f22fc6dd7cbb2a8e0f5a6b858fcc929f3ba/onnx/reference/op_run.py#L232

should I change it to 
alpha = self.alpha if alpha is None else alpha?

Comment (justinchuby) [reply onnx/reference/ops/op_swish.py]: I see. Yes please update to check for None because when alpha is 0 the behavior would be unexpected

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
@isdanni Looks like there is one github comment I cannot resolve to unblock the merge because the commit was gone with the force push. Could you recreate this pull request? You can close the current one and open a new one.
